nuxeo-runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/RuntimeMessageHandlerImpl.java

```diff
@@ -0,0 +1,141 @@
+ * This is interesting to not store several time the same message in case of hot reload.
+ *
+ * @since 9.10
+ */
+public class RuntimeMessageHandlerImpl implements RuntimeMessageHandler, ComponentManager.Listener {
+
+    private final List<Message> messages = new ArrayList<>();
+
+    private ComponentManagerStep step = ComponentManagerStep.ACTIVATION;
+
+    @Override
+    public boolean addWarning(String message) {
+        return messages.add(new Message(step, Level.WARNING, message));
+    }
+
+    @Override
+    public List<String> getWarnings() {
+        return messages.stream().filter(msg -> Level.WARNING.equals(msg.getLevel())).map(Message::getMessage).collect(
+                Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
+    }
+
+    @Override
+    public boolean addError(String message) {
+        return messages.add(new Message(step, Level.SEVERE, message));
+    }
+
+    @Override
+    public List<String> getErrors() {
+        return messages.stream().filter(msg -> Level.SEVERE.equals(msg.getLevel())).map(Message::getMessage).collect(
+                Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
+    }
+
+    @Override
+    public void beforeActivation(ComponentManager mgr) {
+        changeStep(ComponentManagerStep.ACTIVATION);
+    }
+
+    @Override
+    public void beforeStart(ComponentManager mgr, boolean isResume) {
+        changeStep(ComponentManagerStep.START);
+    }
+
+    @Override
+    public void afterStart(ComponentManager mgr, boolean isResume) {
+        changeStep(ComponentManagerStep.RUNNING);
+    }
+
+    @Override
+    public void beforeStop(ComponentManager mgr, boolean isStandby) {
+        changeStep(ComponentManagerStep.STOP);
+    }
+
+    @Override
+    public void beforeDeactivation(ComponentManager mgr) {
+        changeStep(ComponentManagerStep.DEACTIVATION);
+    }
+
+    protected void changeStep(ComponentManagerStep step) {
+        messages.removeIf(msg -> step.equals(msg.getStep()));
+        this.step = step;
+    }
+
+    /**
+     * @since 9.10
+     */
+    protected static class Message {
+
+        private final ComponentManagerStep step;
+
+        private final Level level;
+
+        private final String message;
+
+        public Message(ComponentManagerStep step, Level level, String message) {
+            this.step = step;
+            this.level = level;
+            this.message = message;
+        }
+
+        public ComponentManagerStep getStep() {
+            return step;
+        }
+
+        public Level getLevel() {
+            return level;
+        }
+
+        public String getMessage() {
+            return message;
+        }
+
+    }
+
+    /**
+     * @since 9.10
+     */
+    protected enum ComponentManagerStep {
+
+        ACTIVATION,
+
+        START,
+
+        RUNNING,
+
+        STOP,
+
+        DEACTIVATION
```

Comment:
Consider renaming into "-ING" names: ACTIVATING, STARTING, RUNNING, STOPPING, DEACTIVATING if it makes sense.